src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+
+# default implementations
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: ForwardDiffAD}
+    # TODO: this probably slows down executation quite a bit; exists a better way of doing this?
+    f(θ_) = - vo(alg, q, model, θ_, args...)
+
+    chunk_size = getchunksize(alg)
+    # Set chunk size and do ForwardMode.
+    chunk = ForwardDiff.Chunk(min(length(θ), chunk_size))
+    config = ForwardDiff.GradientConfig(f, θ, chunk)
+    ForwardDiff.gradient!(out, f, θ, config)
+end
+
+# TODO: implement for `Tracker`
+# function grad(vo::ELBO, alg::ADVI, q::MeanField, model::Model, f, autodiff::Val{:backward})
+#     vo_tracked, vo_pullback = Tracker.forward()
+# end
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: TrackerAD}
+    θ_tracked = Tracker.param(θ)
+    y = - vo(alg, q, model, θ_tracked, args...)
+    Tracker.back!(y, 1.0)
+
+    DiffResults.value!(out, Tracker.data(y))
+    DiffResults.gradient!(out, Tracker.grad(θ_tracked))
+end
+
+"""
+    optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad())
+
+Iteratively updates parameters by calling `grad!` and using the given `optimizer` to compute the steps.
+"""
+function optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad()) where AD
+    # TODO: this default constructor for `optimizer` is probably dangerous because of the `acc` field being an `IdDict`!!!
+
+    # TODO: should we always assume `samples_per_step` and `max_iters` for all algos?
+    alg_name = alg_str(alg)
+    samples_per_step = alg.samples_per_step
+    max_iters = alg.max_iters
+
+    # number of previous gradients to use to compute `s` in adaGrad
+    stepsize_num_prev = 10
+    
+    # setup
+    # var_info = Turing.VarInfo()
+    # model(var_info, Turing.SampleFromUniform())
+    # num_params = size(var_info.vals, 1)
+    num_params = length(q)
+
+    # # buffer
+    # θ = zeros(2 * num_params)
+
+    # HACK: re-use previous gradient `acc` if equal in value
+    # Can cause issues if two entries have idenitical values
+    if θ ∉ keys(optimizer.acc)
+        vs = [v for v ∈ keys(optimizer.acc)]
+        idx = findfirst(w -> vcat(q.μ, q.ω) == w, vs)
+        if idx != nothing
+            @info "[$alg_name] Re-using previous optimizer accumulator"
+            θ .= vs[idx]
+        end
+    else
+        @info "[$alg_name] Already present in optimizer acc"
+    end
+    
+    diff_result = DiffResults.GradientResult(θ)
+
+    # TODO: in (Blei et al, 2015) TRUNCATED ADAGrad is suggested; this is not available in Flux.Optimise
```

Comment:
Optimizer can also be an argument to pass in.

Reply:
So instead of being a keyword argument, we make it a positional argument? I'm also coming to the same opinion because of what I mentioned before.

Might make sense to let the `vi(model, alg)` choose a default optimizer for the different methods, rather than having a default optimizer in `optimize!`.